### PR TITLE
Aggregation functions

### PR DESCRIPTION
**Pull request by Blacksmoke16**
Add aggregate methods.  `sum, min, max, and avg`.

### CODE REVIEW

**Comment by Blacksmoke16** [comment]
Few updates on this.  There was firstly an issue with the migrator and pg where the `Float64` was being created as a `REAL` column type.  This was causing fields that were set to type `Float64` to return as `Float32`.

The other issue im coming across is that MySQL shard will default to `Int64` and `Float64` on math, including `SUM`.  I made an issue to see if anyone has any clarification/insight.  

https://github.com/crystal-lang/crystal-mysql/issues/67

File: docs/querying.md [hunk 1]
```diff
@@ -51,6 +51,48 @@ Offset is using the QueryBuilder and provides the ability to offset the results.
 ```crystal
 Post.offset(100).limit(50)
 ```
+## Aggregates
+
+**NOTE:** The return type of each aggregate function can differ depending on the database adapter being used.  For example the `sum ` method:
+
+* `MySQL` will return a `Float64` even if it is an integer.
+* `Postgres` and `Sqlite` will return them a `Int64` or `Int32` depending on the size of the integer.
```
Comment (robacarp) [inline docs/querying.md]: Is there a lot of difference between adapters here? Would this documentation be worth specifying explicitly with the example of the #sum method? If so, Perhaps a table showing the return types for each method?

File: src/granite/query/builder.cr [hunk 2]
```diff
@@ -153,6 +153,16 @@ class Granite::Query::Builder(Model)
     assembler.count
   end
 
+  def aggregate(field : String | Symbol, t)
+    assembler.aggregate(field, t).run
+  end
```
Comment (robacarp) [inline src/granite/query/builder.cr]: I'm not sure why this method is here...why not just make the method defined in the for loop below call the assembler directly?

Comment (Blacksmoke16) [reply src/granite/query/builder.cr]: :man_shrugging: Might have just assumed i need it?  I'll take a look tomorrow.

File: docs/querying.md (hunk 1 above)
Comment (Blacksmoke16) [reply docs/querying.md]: Probably would be a good idea. :+1:

**Comment by Blacksmoke16** [comment]
@robacarp It's because the type of the value coming from db will be DB::Any.  If you do not cast it you'd get errors like `undefined method 'to_i' for Bool (compile-time type is (Bool | Float32 | Float64 | Int32 | Int64 | Slice(UInt8) | String | Time | Nil))`.

Although it might be easier to just return the full union and let user deal with it.  Doesn't really matter who does it.  The expected behavior as it is now would make it seem like it would cast/convert the value to the type you give.  Would that be something we'd want/should do?

**Comment by robacarp** [comment]
@Blacksmoke16 After sleeping on it, I think it's reasonable to ask the developer to specify the return type.

File: docs/querying.md (hunk 1 above)
Comment (Blacksmoke16) [reply docs/querying.md]: Might be easier to just point the user at the specs for them to look at.  The table would be quite complex.  I don't know how to easily display the data without making a _HUGE_ table or a bunch of smaller ones, for something that can be figured out pretty quickly looking at the spec.

File: src/granite/query/executors/value.cr [hunk 3]
```diff
@@ -14,7 +14,9 @@ module Granite::Query::Executor
 
       Model.adapter.open do |db|
         db.query_one? @sql, @args do |record_set|
-          return record_set.read.as Scalar
+          val = record_set.read
+          return val.to_f if val.is_a?(PG::Numeric)
```
Comment (drujensen) [inline src/granite/query/executors/value.cr]: unfortunately, `PG::Numeric` will not exist if you don't include the pg driver.

Comment (Blacksmoke16) [reply src/granite/query/executors/value.cr]: :confused: Ah man...

Comment (robacarp) [reply src/granite/query/executors/value.cr]: There are ways around that. Granite can define a bare class so the const is around. [mosquito does that with granite::base](https://github.com/robacarp/mosquito/blob/master/src/external_classes.cr) and it's worked well.

Comment (Blacksmoke16) [reply src/granite/query/executors/value.cr]: That could work.  I'll have to revisit this and try that.

**Pull request closed without merging**